### PR TITLE
Improve Performance of Kennard-Stone, SPXY Samplers

### PR DESCRIPTION
**Pull request by JacksonBurns**
This PR dramatically improves the performance of the Kennard-Stone (and by consequence, the SPXY) sampler.

This new implementation makes a few major changes:
 - the list of indices which are selected already serves an additional purpose: masking the distance matrix to avoid us having to delete from the distance matrix
 - we only re-calculate the minimum distance of the unselected sample to the _new_ sample on each for loop iteration, which effectively reduces the runtime complexity by order N (to O(N^2))

The below code snippet is what I have been using to test, and `astartes` finishes in 3.7 seconds whereas `kennard_stone` takes 11 seconds (this also holds true for other size input arrays than the one here, this is on Python 3.7 with the latest version of kennard-stone installed from PyPI):
```python
import numpy as np
from kennard_stone import train_test_split as kennardstone_train_test_split
from astartes import train_test_split as astartes_train_test_split
example_X = np.random.rand(10_000, 100)
from time import perf_counter
start = perf_counter()
astartes_train_test_split(
    example_X,
    sampler="kennard_stone",
    hopts=dict(
        metric="euclidean",
    ),
)
stop = perf_counter()
print("astartes elapsed time: ", stop - start)
start = perf_counter()
kennardstone_train_test_split(
    example_X,
    metric="euclidean",
)
stop = perf_counter()
print("kennard_stone elapsed time: ", stop - start)
```

### CODE REVIEW

**Review by kspieks — APPROVED** [review]
Thanks for this update! LGTM
